diff --git a/board/kosagi/novena/novena_spl.c b/board/kosagi/novena/novena_spl.c
index 008418b018..4f2a9caeb2 100644
--- a/board/kosagi/novena/novena_spl.c
+++ b/board/kosagi/novena/novena_spl.c
@@ -386,7 +386,7 @@ static inline void novena_spl_setup_iomux_video(void) {}
  */
 #ifdef CONFIG_FSL_ESDHC_IMX
 static struct fsl_esdhc_cfg usdhc_cfg = {
-	USDHC3_BASE_ADDR, 0, 4
+	USDHC2_BASE_ADDR, 0, 4
 };
 
 int board_mmc_getcd(struct mmc *mmc)
@@ -397,7 +397,7 @@ int board_mmc_getcd(struct mmc *mmc)
 
 int board_mmc_init(struct bd_info *bis)
 {
-	usdhc_cfg.sdhc_clk = mxc_get_clock(MXC_ESDHC3_CLK);
+	usdhc_cfg.sdhc_clk = mxc_get_clock(MXC_ESDHC2_CLK);
 	return fsl_esdhc_initialize(bis, &usdhc_cfg);
 }
 #endif
