diff --git a/include/configs/novena.h b/include/configs/novena.h
index 12fb759afe..ca59a57b75 100644
--- a/include/configs/novena.h
+++ b/include/configs/novena.h
@@ -56,8 +56,6 @@
 	"initrd_high=0xffffffff\0"					\
 	"consdev=ttymxc1\0"						\
 	"baudrate=115200\0"						\
-	"bootdev=/dev/mmcblk0p1\0"					\
-	"rootdev=/dev/mmcblk0p2\0"					\
 	"netdev=eth0\0"							\
 	"kernel_addr_r="__stringify(CONFIG_SYS_LOAD_ADDR)"\0"		\
 	"pxefile_addr_r="__stringify(CONFIG_SYS_LOAD_ADDR)"\0"		\
@@ -77,43 +75,19 @@
 	"addmisc="							\
 		"setenv bootargs ${bootargs} ${miscargs}\0"		\
 	"addargs=run addcons addmisc\0"					\
-	"mmcload="							\
-		"mmc rescan ; "						\
-		"ext4load mmc 0:1 ${kernel_addr_r} ${bootfile}\0"	\
 	"netload="							\
 		"tftp ${kernel_addr_r} ${hostname}/${bootfile}\0"	\
 	"miscargs=nohlt panic=1\0"					\
-	"mmcargs=setenv bootargs root=${rootdev} rw rootwait\0"		\
 	"nfsargs="							\
 		"setenv bootargs root=/dev/nfs rw "			\
 			"nfsroot=${serverip}:${rootpath},v3,tcp\0"	\
-	"mmc_mmc="							\
-		"run mmcload mmcargs addargs ; "			\
-		"bootm ${kernel_addr_r}\0"				\
-	"mmc_nfs="							\
-		"run mmcload nfsargs addip addargs ; "			\
-		"bootm ${kernel_addr_r}\0"				\
-	"net_mmc="							\
-		"run netload mmcargs addargs ; "			\
-		"bootm ${kernel_addr_r}\0"				\
 	"net_nfs="							\
 		"run netload nfsargs addip addargs ; "			\
 		"bootm ${kernel_addr_r}\0"				\
-	"update_sd_spl_filename=SPL\0"					\
-	"update_sd_uboot_filename=u-boot.img\0"				\
-	"update_sd_firmware="	/* Update the SD firmware partition */	\
-		"if mmc rescan ; then "					\
-		"if dhcp ${update_sd_spl_filename} ; then "		\
-		"mmc write ${loadaddr} 2 0x200 ; "			\
-		"fi ; "							\
-		"if dhcp ${update_sd_uboot_filename} ; then "		\
-		"fatwrite mmc 0:1 ${loadaddr} u-boot.img ${filesize} ; "\
-		"fi ; "							\
-		"fi\0"							\
 	BOOTENV
 
 #define BOOT_TARGET_DEVICES(func) \
-	func(MMC, mmc, 0) \
+	func(MMC, mmc, 1) \
 	func(USB, usb, 0) \
 	func(SATA, sata, 0) \
 	func(PXE, pxe, na) \
